### ISSUE TITLE
Deliverable: Full documentation

### ISSUE THREAD

**Issue by ronaldtse** [ietf-ribose/github-migration-project#12]
1. List of all page redirects that are required to ensure that external references to the existing Trac wikis are still functional. IETF will use this list to implement a redirect to the new page.
2. List of every Trac Wiki page migrated and its new target in GitHub Pages
3. List of every Trac Wiki page not migrated and why it was not moved.
4. Mappings from Subversion revision numbers to Git commit ids
5. Mappings from Trac ticket numbers to GitHub issue numbers.

UPDATED: 2021-11-22

**Comment by HassanAkbar** [comment]
@ronaldtse Need some clarification on the following:

1. From what I understand this means List of all the external links in GitHub Issues created by migrating tickets from trac. Is that right or is there something else?
2. Can you please elaborate this.
3. Can you please elaborate this.
4. Clear. We will be using revmap for this.
5. Clear.

**Comment by ronaldtse** [comment]
@HassanAkbar clarified in the OP. Does that make sense?

**Comment by HassanAkbar** [comment]
Yes, Thank you.

**Comment by HassanAkbar** [comment]
@ronaldtse 

- I have done some research and found [this script](https://github.com/ggainey/tracwiki2githubwiki) for `Trac -> Github Wiki` migration. We can tweak this according to our needs. [Here](https://github.com/hassanakbar4/foobar/wiki) are the migrated `wikis` for `ietfdb`.
- The script also copy the text into files with their version histories from trac.
- But this script does not change the formatting of the text (pushes wikis with` trac wiki formatting` to Github). For that we can call the converter from `tractive` to convert `trac wiki` format to `markdown` format.
- Attachments are currently missing we will need to add the `base url` of the repo for that. But the script has the ability to handle attachments if we provide `base url` to it.

**Comment by ronaldtse** [comment]
@HassanAkbar let's re-implement the conversion mechanism used by the [script](https://github.com/ggainey/tracwiki2githubwiki) (apparently by Red Hat) into Tractive directly in Ruby. That would enable easier migration for people who use Tractive in the future.

**Comment by ronaldtse** [comment]
Wiki issue here: https://github.com/ietf-ribose/tractive/issues/63

**Comment by HassanAkbar** [comment]
> @HassanAkbar let's re-implement the conversion mechanism used by the [script](https://github.com/ggainey/tracwiki2githubwiki) (apparently by Red Hat) into Tractive directly in Ruby. That would enable easier migration for people who use Tractive in the future.

Agreed @ronaldtse. I will get it started.

**Comment by HassanAkbar** [comment]
@rjsparks In `ietfdb` there are 2 different wikis named `Sandbox` and `SandBox`, But GitHub does now allow to make 2 different wikis with same name so we need to rename or remove one of these. 

<img width="250" alt="Screenshot 2021-12-08 at 5 28 36 PM" src="https://user-images.githubusercontent.com/5301572/145209069-46470399-c667-4c89-b977-4016b71a5f12.png">

What do you suggest we should do here?

_Originally posted by @HassanAkbar in https://github.com/ietf-ribose/tractive/issues/63#issuecomment-988774430_

**Comment by rjsparks** [comment]
I renamed one of the pages to disambiguate them.
`The page SandBox has been renamed to EditingSandBox.`
And is viewable now at https://trac.ietf.org/trac/ietfdb/wiki/EditingSandBox.

**Comment by HassanAkbar** [comment]
The sample wikis have been migrated to the following repos:
- datatracker -> https://github.com/ietf-svn-conversion/ietfdb/wiki
- xml2rfc -> https://github.com/ietf-svn-conversion/xml2rfc/wiki

@rjsparks I also want to confirm that what should we do with wikis for repos like `datatracker` and `mailarch`? 
Because both these repos have same trac instance but are migrated to different `Git repos`. Same goes for `xml2rfc`, `xml2rfc-bibxml`, `xml2rfc-website`, and `vocabulary_design_team_2013_2017`, all of these have `xml2rfc` as trac instance.

cc: @ronaldtse

**Comment by rjsparks** [comment]
The entire ietfdb wiki should move to the datatracker github instance. The entire xml2rfc wiki should move to the xml2rfc github instance.

**Comment by HassanAkbar** [comment]
@rjsparks The sample wikis with attachments links have been migrated to the following repos:

- datatracker -> https://github.com/ietf-svn-conversion/datatracker-updated/wiki
- xml2rfc -> https://github.com/ietf-svn-conversion/xml2rfc-updated/wiki

**Comment by rjsparks** [comment]
If it's easy, mapping WikiStart to Home would match what trac is doing. If it's not easy, it's a tiny thing to do manually afterwards.

**Comment by rjsparks** [comment]
First thoughts - there are a lot of places where Trac WikiFormat isn't being converted to something reasonable yet. I hope the conversion script can deal with more of these than it currently does. See https://github.com/ietf-svn-conversion/datatracker-updated/wiki/IETF110.1Sprint 

Note that it didn't honor the #!html or #!comment WikiFormat commands - just translating them to codeblocks is better than losing them, but showing the comments is unfortunate, and surely there's a github markdown concept to map it into?

More importantly the internal links from this page to other wiki pages didn't get built - see`If so, please sign up on the [wiki:IETF110.1SprintSignUp] page.` for example.

**Comment by rjsparks** [comment]
And then there are references to the trac instance we're migrating away from that haven't been migrated forward.
It's not going to be easy, and probably not even possible, to automate replacing them, but perhaps we could flag them,  turn them into something that are no longer links (but preserve what the old link was), or something else, because we _know_ that they aren't going to work post migration.

For examples see:
```
If you're looking for a short task to work on, look for rows marked "easy" in this query of Datatracker tickets. Please consider the higher priority items first.

If there's something you hope the sprint will address that's not on that list, please enter it as a new ticket using http://trac.ietf.org/trac/ietfdb/newticket
```
The link at `"easy" in this query of Datatracker tickets` currently points to http://trac.ietf.org/trac/ietfdb/query?status=assigned&status=new&status=reopened&status=accepted&keywords=~sprint+easy&group=priority&max=0&col=id&col=summary&col=keywords&col=owner&col=component&desc=1&order=id

**Comment by jennifer-richards** [comment]
It looks like the WikiWords did not turn into links much if at all - eg, the list of links at the top of WikiStart are just text.

**Comment by jennifer-richards** [comment]
It looks like things that should be `treated as code` appear as [source:this is not code] instead. E.g., https://github.com/ietf-svn-conversion/ietfdb/wiki/CodeRepository in the second paragraph of the top section

**Comment by rjsparks** [comment]
> It looks like the WikiWords did not turn into links much if at all - eg, the list of links at the top of WikiStart are just text.

This is a very big deal - tracwiki formating has an implicit rule that any CamelCased word is a link to an internal wiki page with that title. The conversion tool really must implement adding links for those words to the associated github wiki page.

Note that there is a convention that !CamelCasedWord told the trac renderer not to create a link. It would be a nice to have feature for the converter to drop the ! in such cases and not create a link.

**Comment by rjsparks** [comment]
Tables don't convert : see https://github.com/ietf-svn-conversion/ietfdb/wiki/DatatrackerDrafts

**Comment by rjsparks** [comment]
On inspection, the source wiki for xml2rfc is essentially unused, and not worth preserving. We can drop migrating that one.

**Comment by HassanAkbar** [comment]
I fixed the above-mentioned issues in the wikis, here is the latest wiki for `datatracker` -> https://github.com/ietf-svn-conversion/datatracker/wiki

@rjsparks  can you give it a review ?

Here is the list of trac instance URLs in wikis:

- [IETF110Sprint](https://github.com/ietf-svn-conversion/datatracker/wiki/IETF110Sprint):
  - http://trac.ietf.org/trac/ietfdb/query?status=assigned&status=new&status=reopened&status=accepted&keywords=~sprint+easy&group=priority&max=0&col=id&col=summary&col=keywords&col=owner&col=component&desc=1&order=id
  - http://trac.ietf.org/trac/ietfdb/newticket
  - http://trac.ietf.org/trac/ietfdb/newticket?keywords=sprint&type=enhancement&priority=n/a&version=&component=Other

- [IETF110.1Sprint](https://github.com/ietf-svn-conversion/datatracker/wiki/IETF110.1Sprint):
  - http://trac.ietf.org/trac/ietfdb/query?status=assigned&status=new&status=reopened&status=accepted&keywords=~sprint+easy&group=priority&max=0&col=id&col=summary&col=keywords&col=owner&col=component&desc=1&order=id
  - http://trac.ietf.org/trac/ietfdb/newticket
  - http://trac.ietf.org/trac/ietfdb/newticket?keywords=sprint&type=enhancement&priority=n/a&version=&component=Other

**Comment by rjsparks** [comment]
things at /datatracker/wiki look mostly as expected now. The "Testing" link at the top of wikistart didn't translate, but it's not a link in trac either (flaw in the way we were using Trac).

**Comment by rjsparks** [comment]
I see that https://github.com/ietf-svn-conversion/datatracker/issues is empty?
Is there an instance of the conversion run that has everything that's intended to be in place?
I'd like to look there to see if I can suggest automating some of the ticket searches.

The newticket links _could_ be algorithmically changed to point to "create an issue" (or "start a discussion"). But since we're going to have change the process around that anyhow, it may be better to leave those links so we can search for them and address the surrounding text as we fix them.

**Comment by HassanAkbar** [comment]
> I see that https://github.com/ietf-svn-conversion/datatracker/issues is empty?
> Is there an instance of the conversion run that has everything that's intended to be in place?
> I'd like to look there to see if I can suggest automating some of the ticket searches.

@rjsparks Migration of issues in the repo is in progress, will update once it's done.

**Comment by HassanAkbar** [comment]
@rjsparks Here are the migrated repos with wikis and issues:

- datatracker -> https://github.com/ietf-svn-conversion/datatracker
- mailarchive -> https://github.com/ietf-svn-conversion/mailarch
- xml2rfc -> https://github.com/ietf-svn-conversion/xml2rfc

**Comment by HassanAkbar** [comment]
The migration has been run on the following repos.
- mailreplay -> https://github.com/ietf-svn-conversion/mailreplay
- postconfirm -> https://github.com/ietf-svn-conversion/postconfirm
- postfind -> https://github.com/ietf-svn-conversion/postfind
- tsvwg -> https://github.com/ietf-svn-conversion/tsvwg
- emailcore -> https://github.com/ietf-svn-conversion/emailcore
- dmarc -> https://github.com/ietf-svn-conversion/dmarc

**Comment by ronaldtse** [comment]
@rjsparks I believe the latest batch of migrations are ready to go. Could you help review and confirm, then we can hand over those smaller repos? Thanks!

**Comment by rjsparks** [comment]
@ronaldtse 
For (mailreplay, postconfirm, postfind) - if we find no issues in reviewing today/tomorrow, should we take those as final, or do you need to make a final run?

The other three (tsvwg, emailcore, dmarc) will require re-coordination with the respective group chairs, and emailcore in particular may require another 'final' run. Will advise when I know.

**Comment by rjsparks** [comment]
@ronaldtse - reping: I am ready to accept mailreplay/postconfirm/postfind, Is what's there the final run, or do you need to make one last one? (There have been no changes in the source repositories).

I plan to transfer those repos to the ietf-tools org unless you recommend a different mechanism.

**Comment by ronaldtse** [comment]
@rjsparks sorry for the delay in replying. For (mailreplay, postconfirm, postfind), there has been no changes and can be now transferred. Please feel free to proceed!

(We just need to comment out the GHA jobs after.)

**Comment by HassanAkbar** [comment]
@rjsparks We need to finalize one `repo` that will store all the images for `trac tickets and wikis`.
The repo should be `public` and in `IETF` Github organization.
If we will use an internal repo and transfer ownership later all the image paths will break.
So, its important to finalize a repo for this purpose prior to running final migration.

Cc: @ronaldtse

**Comment by rjsparks** [comment]
@HassanAkbar -  detail: Where you say `IETF` Github organization, I think you should be saying `IETF Tools`. That is - things will land at https://github.com/ietf-tools/, not https://github.com/ietf/. Does that matter to your code or the process we're about to go through?

This is one repo that will be used for _all_ of the transferred repositories, or is it one image repo per repository to transfer? (I'm curious why these don't end up as part of the individual repositories)?

**Comment by rjsparks** [comment]
@HassanAkbar - The (mailreplay, postconfirm, postfind) repos have no trac content at all. Do they need to be otherwise finalized?

**Comment by HassanAkbar** [comment]
> @HassanAkbar - The (mailreplay, postconfirm, postfind) repos have no trac content at all. Do they need to be otherwise finalized?

@rjsparks Yes, they are ready. 

For repos containing trac content, we will be needing a repo to store image.

**Comment by ronaldtse** [comment]
@rjsparks the challenge with images attached to GitHub issues (from Trac), is that we cannot migrate them to the issues themselves — GitHub does not provide an API for uploading attachments (it only works on the GitHub site and apps). Currently, there is no method for create a GitHub attachment using automation. 

This means we need to migrate the images somewhere, and the easiest/most accessible option is to keep them in a new repo which sole usage is to store the migrated images, which can be displayed in the migrated issues. 

@HassanAkbar can we just create a repo called “trac-attachments” and we migrate all the attachments there?

Thanks!

**Comment by rjsparks** [comment]
Thinking through what the future may bring. I'm hearing that we will have one repo in ietf-tools that is a bucket for trac images for any of the repos we convert with this project - if that's off base (particularly if it's one image repo per converted repo) let me know.

Some of these converted repos, particularly the working group issue tracker repositories (tsvwg, emailcore, dmarc) are going to start out in organizations other than ietf-tools when we accept them, and will end up with a different set of owners. Will what you've set up work if the image bucket is not in the same organization as the converted repository?

Having a single repo like this puts a strange burden on us - we could easily lose track of the relying repositories. Could they, in the future, fork the image bucket and stop relying on us? Or are paths somehow buried in the result of what you're doing that would prevent that?

Having a converted repository be self-contained would have been much better - did you explore putting these attachements into each converted repostitory at a path that was not likely to exist instead?

**Comment by rjsparks** [comment]
@HassanAkbar , @ronaldtse : I'm going to transfer (mailreplay, postconfirm, postfind) in a few minutes.

**Comment by HassanAkbar** [comment]
> Thinking through what the future may bring. I'm hearing that we will have one repo in ietf-tools that is a bucket for trac images for any of the repos we convert with this project - if that's off base (particularly if it's one image repo per converted repo) let me know.
> 
> Some of these converted repos, particularly the working group issue tracker repositories (tsvwg, emailcore, dmarc) are going to start out in organizations other than ietf-tools when we accept them, and will end up with a different set of owners. Will what you've set up work if the image bucket is not in the same organization as the converted repository?
> 
> Having a single repo like this puts a strange burden on us - we could easily lose track of the relying repositories. Could they, in the future, fork the image bucket and stop relying on us? Or are paths somehow buried in the result of what you're doing that would prevent that?
> 
> Having a converted repository be self-contained would have been much better - did you explore putting these attachements into each converted repostitory at a path that was not likely to exist instead?

@rjsparks Actually, we can go with any of the following options:
1. Keep all attachments in one repo.
2. Keep separate attachments repo for every repo.
3. Create a separate directory in every repo named `attachments` and save files in there. This will also remove any dependency. Github allows using [relative paths](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#images).


We can go with the 3rd option as it removes any dependency. 
Only `datatracker`, `mailarch` and `xml2rfc` contain attachments. The size of `attachments` directory will be between 20-26 MB.

cc: @ronaldtse

**Comment by rjsparks** [comment]
Yes, option 3 please. a top-level directory that has a very obvious name like 'svntrac-converted-attachments' should work fine.

I take it from this that the (tsvwg, emailcore, and dmarc) repositories are then ready to be transferred now, correct? (@HassanAkbar). (btw, let me know if all this extra @ signalling in the issue is helpful right now or if it unnecessary)

**Comment by HassanAkbar** [comment]
>I take it from this that the (tsvwg, emailcore, and dmarc) repositories are then ready to be transferred now, correct? 

If there are no new tickets after the last migration which was about 4 days ago then yes (tsvwg, emailcore, and dmarc) repos are ready.

>(btw, let me know if all this extra @ signalling in the issue is helpful right now or if it unnecessary)

You are right it's unnecessary I guess 😄

**Comment by rjsparks** [comment]
I've transfered the three code repositories.

**Comment by rjsparks** [comment]
I've transferred dmarc

**Comment by rjsparks** [comment]
@HassanAkbar : emailcore needs to be rerun (there were updates 20 hrs ago). Please let me know when done.

**Comment by ronaldtse** [comment]
> Yes, option 3 please. a top-level directory that has a very obvious name like 'svntrac-converted-attachments' should work fine.

@rjsparks I would like to confirm this because the attachments will pollute the history of these repos (datatracker, mailarch and xml2rfc). They will also add a size burden to those particular repos since Git never removes anything -- if you wanted to remove those attachments in the future, a repository rewrite (force push to change history) is necessary.

**Comment by rjsparks** [comment]
Understood.

Out of curiosity, how big _is_ this attachment auxiliary   right now? I mean, they _are_ history - so I think pollute is not quite the right word, but carrying the weight of that history has a cost.

I still think severability of the repositories is a feature we would want in the long run, and having a single auxiliary attachment repository complicates that. I think then, we may need to compromise on #2.

**Comment by NGPixel** [comment]
I'm not too fond of having issues attachments in the git repository itself (option #3). As @ronaldtse mentionned, it will simply pollute the code base with data that isn't code at all and can't be easily removed afterwards. Also, any clone or fork will result in copying all these files which would use extra bandwidth and unnecessary disk space for everyone.

Right now, if you add an image / file to a GitHub issue, it is not stored in the git repo itself, so I don't think we should have old attachments in there either.

A single repository with all attachments would be much better IMO (option #1). They could be stored inside subdirectories matching the name of the repo they reference. (e.g. "datatracker/1234-1.jpg" for issue #1234 from datatracker project).

That would eliminate the need for duplicate repositories for all projects (option #2), which would again pollute the organization with repositories that will essentially become read-only...

**Comment by ronaldtse** [comment]
I echo @NGPixel ‘s comments. I think the correct resolution here is to use Option 1 (single external repository for attachments), and then manually migrate those attachments back to the tickets, repo by repo. This way we can slowly wean off the attachment repo and dispose it then. 

This depends on how many attachments there are so I’ll need to double check the amount of attachments…

**Comment by ronaldtse** [comment]
I've done a preliminary search using a older Trac dataset (on mobile right now), and it turns out there are ~500 attachments across these repositories (mailarch is included in the `ietfdb` Trac instance). Size is not an issue but the number of files means manual migration will take a while.

```
find xml2rfc/files/attachments -type f | wc -l
     350
Size: 25M
     
find ietfdb/files/attachments -type f | wc -l
     178
Size: 19M
```

I suggest we use Option 1 for now, and find a way to manually migrate those ~500 attachments (either we get someone to do this, or get some intern to do so), or wait until GitHub provides an attachment upload API to automatically migrate those attachments.

**Comment by rjsparks** [comment]
That the issues (even with their attachments) aren't part of the repository at github has been a _problem_ for many of the working groups using it - Martin's draft tool goes to great lengths to create a branch that has a copy of the issues, so, again, I don't think pollute is the right word.

I'll relent and accept path #1, with the understanding and expectation that it will only have content from datatracker, mailarchive, and xml2rfc.

**Comment by HassanAkbar** [comment]
Btw if the repo is renamed/transferred to another organization, Github maintains an internal mapping from the old URL -> new URL (it breaks when you create a new repo with an old name).

So, if the attachment repo is transferred it might not create a problem until we create another repo with the same name.
Any preference should I create this repo in `ietf-svn-conversion` or do you want a specific repo in another organization to be used for the final migration?

**Comment by HassanAkbar** [comment]
> @HassanAkbar : emailcore needs to be rerun (there were updates 20 hrs ago). Please let me know when done.

 Migration for emailcore and tsvwg are run again. Please verify.

**Comment by rjsparks** [comment]
> do you want a specific repo in another organization to be used for the final migration?

We can create it directly in the ietf-tools organization if that simplifies things. What have you been naming it?

**Comment by HassanAkbar** [comment]
I have added all attachments to [svntrac-converted-attachments repo](https://github.com/ietf-svn-conversion/svntrac-converted-attachments). Can you move it to the `ietf-tools` organization ?

**Comment by rjsparks** [comment]
done

**Comment by rjsparks** [comment]
emailcore and tsvwg have been transferred - all the simple repos are now accepted.

@HassanAkbar we'd like to move xml2rfc next, and this week if possible - could you give us a final candidate to transfer?

**Comment by rjsparks** [comment]
@HassanAkbar - it occurs to me that we have a lot of activity going into the xm2rfc repository today. I'll send a note when we get to a quiescent point, and you can start a full run for it?

**Comment by HassanAkbar** [comment]
> @HassanAkbar - it occurs to me that we have a lot of activity going into the xm2rfc repository today. I'll send a note when we get to a quiescent point, and you can start a full run for it?

Sounds good 👍

**Comment by rjsparks** [comment]
@HassanAkbar - we are done touching xml2rfc for today and will consider things to be in a freeze state.

**Comment by HassanAkbar** [comment]
> @HassanAkbar - we are done touching xml2rfc for today and will consider things to be in a freeze state.

@rjsparks I am feeling a bit under the weather today. But I will try to get it done as soon I get back (hopefully tomorrow).

**Comment by rjsparks** [comment]
> I am feeling a bit under the weather today. But I will try to get it done as soon I get back (hopefully tomorrow).

Feel better soon.

**Comment by HassanAkbar** [comment]
I am ready to run the migration of the following repos

- datatracker
- mailarch
- vocabulary_design_team_2013_2017
- xml2rfc
- xml2rfc-bibxml
- xml2rfc-website

But it will take some time around 6 to 7 hours to complete. Should I start the migration now or should I run it on the weekend?

**Comment by HassanAkbar** [comment]
I am ready to run the migration of the following repos:

- datatracker
- mailarch
- vocabulary_design_team_2013_2017
- xml2rfc
- xml2rfc-bibxml
- xml2rfc-website

But it will take some time around 6 to 7 hours to complete. Should I start the migration now or should I run it on the weekend?

**Comment by rjsparks** [comment]
Start it now please. We are planning to take the xml2rfc* and vocab*repos as final, assuming we don't see issues. We will need to run the datatracker and mailarch again in the future.

**Comment by HassanAkbar** [comment]
The migration is now in progress will be completed in a couple of hours.

**Comment by HassanAkbar** [comment]
The following repos are ready:

- xml2rfc -> https://github.com/ietf-svn-conversion/xml2rfc
- vocabulary_design_team_2013_2017 -> https://github.com/ietf-svn-conversion/vocabulary_design_team_2013_2017
- xml2rfc-bibxml -> https://github.com/ietf-svn-conversion/xml2rfc-bibxml
- xml2rfc-website -> https://github.com/ietf-svn-conversion/xml2rfc-website

**Comment by rjsparks** [comment]
@HassanAkbar  - the xml2rfc repository issues don't look right this time. They stop at 446, and they should be going to 708 (709 if you were to update the source and run it again). What's happened?

**Comment by HassanAkbar** [comment]
I have rerun the migration, Please verify.

@ronaldtse  The issue has been resolved in this [PR](https://github.com/ietf-ribose/tractive/pull/99).

**Comment by ronald** [comment]
@ronaldtse

**Comment by rjsparks** [comment]
@HassanAkbar I see one thing to ask about. Merges seem to have been processed differently here than they were in the datatracker project. In the xml2rfc repository, the merges from kesara and jennifer-richards aren't captured as commits into main, and they don't appear in the contributors on the repository's main page. Their contributions to the datatracker, merged the same way in svn as far as I can tell so far, _are_ reflected as commits from them in main in that project (and they show up as contributors).

For xml2rfc, I don't think it's worth slowing down for another run (unless you see an obvious thing that broke the conversion otherwise). But I would like to understand why the conversions are different, and make sure that the behavior of the conversion for the datatracker doesn't change when we make its final run.

**Comment by ronaldtse** [comment]
(Thanks @ronald!)

> In the xml2rfc repository, the merges from kesara and jennifer-richards aren't captured as commits into main, and they don't appear in the contributors on the repository's main page.

@rjsparks from my experience I believe GitHub does not count merges as contributions.

This is what GitHub says:
> Contributions to main, excluding merge commits and bot accounts

https://github.com/ietf-svn-conversion/xml2rfc/graphs/contributors
<img width="857" alt="Screenshot 2022-02-08 at 11 52 37 PM" src="https://user-images.githubusercontent.com/11865/153024345-5e7e4251-434b-4f8b-93af-15fe1b3105f1.png">

**Comment by rjsparks** [comment]
So, I went down the same path as you, but what that means is that github doesn't count the top level "merge commit" when a pr is merged and not squashed. the constituent commits _are_ counted. That's how contributors come in. A similar thing happens with squashes.

Can you tell me what the essential difference between this pair of commits is?

**Comment by rjsparks** [comment]
Compare https://trac.ietf.org/trac/ietfdb/changeset/19901/ and https://trac.ietf.org/trac/xml2rfc/changeset/4093/
and their analogues at:
https://github.com/ietf-svn-conversion/datatracker/commit/ac724096749f7bf638016eca9dd64136a779ba24 and https://github.com/ietf-svn-conversion/xml2rfc/commit/8b293ea14870c4111b4cf3a3f379c0af201641dc

**Comment by rjsparks** [comment]
And, to help focus on my question. Note the commits showing from kesara and jennifer here: 
https://github.com/ietf-svn-conversion/datatracker/commits/main (scroll down a bit to see jennifer-richards)
vs not appearing here:
https://github.com/ietf-svn-conversion/xml2rfc/commits/main

What happened to make the processing different?

**Comment by jennifer-richards** [comment]
One difference I see is that the `svn:mergeinfo` property is updated on `trunk/cli/.` in xml2rfc vs `trunk/.` in datatracker. Perhaps the mechanism for linking merge commits to their origins is not seeing the data in xml2rfc because it's a layer deeper in the tree?

**Comment by rjsparks** [comment]
Good catch - I suspect that's exactly the difference. I don't think we should slow down to try to deal with that kind of merge, even if it was common for xml2rfc. Let's move ahead with this conversion (I'll work on transfers in the next hour).

**Comment by ronaldtse** [comment]
Thanks @jennifer-richards and @rjsparks for the investigation. It's possible that the `reposurgeon` tool we used was not able to migrate the `svn:mergeinfo` with the right information, which alluded GitHub's contributor counting mechanism.

**Comment by rjsparks** [comment]
@HassanAkbar: We could use the list of mappings from svn commit number to github commit hash for the xml2rfc repo now. Where can we find it?

**Comment by HassanAkbar** [comment]
Here is the link to svn number to Github commit hash file for xml2rfc -> https://github.com/ietf-ribose/svn-git-maps/blob/929757c2465133ebe9fa4b19fa6b3865d49dce1a/xml2rfc-revmap.txt

**Comment by rjsparks** [comment]
@HassanAkbar @ronaldtse : We are getting close to the point where we are ready for the final datatracker/mailarchive conversion run. We'll have to schedule that one more firmly as the frequency of ticket and code changes is higher. But I anticipate this happening next week or the week after. I note that the github-svn-bot api token expired and am talking with Ron about how to get a new one to you. Other than that, is anything else needed before we schedule this final run?

**Comment by ronaldtse** [comment]
@rjsparks thanks, I've received the updated token and updated the token secret in this GitHub organization.

I'll let @HassanAkbar confirm on what else is needed but I know we need to address #26.

**Comment by rjsparks** [comment]
If it can be done quickly, it would be quite nice to address #27 if at all possible.

**Comment by ronaldtse** [comment]
@rjsparks we should address #27 prior to migration.

**Comment by HassanAkbar** [comment]
> @HassanAkbar @ronaldtse : We are getting close to the point where we are ready for the final datatracker/mailarchive conversion run. We'll have to schedule that one more firmly as the frequency of ticket and code changes is higher. But I anticipate this happening next week or the week after. I note that the github-svn-bot api token expired and am talking with Ron about how to get a new one to you. Other than that, is anything else needed before we schedule this final run?

I think we have everything for the migration, we only need to resolve #26 before migration.

**Comment by rjsparks** [comment]
@HassanAkbar - if things are set up, could we run a next-to-last full conversion of ietfdb into the repositories we're bulding from it  (including the wiki migrations) this weekend so we could re-verify before we declare a freez and make a final run sometime during the week next week? If those times aren't realistic, let me know when it would be reasonable to do so?

**Comment by HassanAkbar** [comment]
I've started the migration for `datatracker` and `mailarch`, It will be done hopefully by tomorrow. I will let you know when its done.

**Comment by HassanAkbar** [comment]
Migration is done for the following repos

- datatracker -> https://github.com/ietf-svn-conversion/datatracker
- mailarch -> https://github.com/ietf-svn-conversion/mailarch

**Comment by rjsparks** [comment]
Thanks! We're making final checks - if all goes well, we'll freeze (for a certain value of freeze) and ask for the final conversion either US Wed or Thu next week.

**Comment by jennifer-richards** [comment]
There seem to be a few merge commits that are not linked to their origin. E.g., ecf768d5 (aka r19932) merged b4d07e1 (r19925) but does not list this as a parent. Appears to affect older work as well - e.g., 8dc76252 (r18947) merged 497a453 (r18938).

This makes it a bit hard to identify whether a branch has been abandoned or merged back, though this can be found without too much trouble by searching for the `Legacy-Id`.  I've only found a handful of instances of this - most merges seem to have correct linkage.

**Comment by rjsparks** [comment]
@HassanAkbar @ronaldtse Is this discrepancy easy to explain? It's not immediately obvious, and I'd like to know it's not the tip of an iceburg.

**Comment by rjsparks** [comment]
Also at https://github.com/ietf-svn-conversion/mailarch/issues/1 - I thought the conversion was going to remove these placeholder issues (many thousands of them for this repo) at the end of the run. Did we decide otherwise somewhere along the way?

**Comment by HassanAkbar** [comment]
> Also at [ietf-svn-conversion/mailarch#1](https://github.com/ietf-svn-conversion/mailarch/issues/1) - I thought the conversion was going to remove these placeholder issues (many thousands of them for this repo) at the end of the run. Did we decide otherwise somewhere along the way?

The flag was missing in the script, I have added that. Now all the placeholder issues have been deleted.

> @HassanAkbar @ronaldtse Is this discrepancy easy to explain? It's not immediately obvious, and I'd like to know it's not the tip of an iceburg.

Looking into this.

**Comment by ronaldtse** [comment]
>  I thought the conversion was going to remove these placeholder issues (many thousands of them for this repo) at the end of the run. 

Yes this was implemented in https://github.com/ietf-ribose/tractive/pull/89 .

**Comment by HassanAkbar** [comment]
> There seem to be a few merge commits that are not linked to their origin. E.g., ecf768d5 (aka r19932) merged b4d07e1 (r19925) but does not list this as a parent. Appears to affect older work as well - e.g., 8dc76252 (r18947) merged 497a453 (r18938).

I looked into it and it turns out that [ecf768d5 (r19932)](https://github.com/ietf-svn-conversion/datatracker/commit/ecf768d544429b01754df2ab85d258d82818396b) is merged into `main` while [b4d07e1 (r19925)](https://github.com/ietf-svn-conversion/datatracker/commit/b4d07e11519d77523216bc8817058b0facd2bd41) is merged into `personal/jennifer/7.44.1.dev0`.
<img width="1368" alt="Screenshot 2022-02-24 at 6 28 31 PM" src="https://user-images.githubusercontent.com/5301572/155533209-86d17483-bfc8-4e26-8801-0bec40cabdae.png">
<img width="1368" alt="Screenshot 2022-02-24 at 6 29 27 PM" src="https://user-images.githubusercontent.com/5301572/155533275-4ec95499-e523-4e97-8beb-af10af81d64c.png">


But if we look at another example like [9be597db (r19956)](https://github.com/ietf-svn-conversion/datatracker/commit/9be597db615463b94c8a1d0fd00da6ace9cf4693) which merged [bc8155f1 (r19947)](https://github.com/ietf-svn-conversion/datatracker/commit/bc8155f197cb06415983016defc988154cb6c4a5). We can see that both commits are merged into `main` and the parents are set correctly.
<img width="1368" alt="Screenshot 2022-02-24 at 6 35 26 PM" src="https://user-images.githubusercontent.com/5301572/155534986-dc96a040-2a16-46e8-aaa2-3aeb5cb0d378.png">
<img width="1368" alt="Screenshot 2022-02-24 at 6 39 57 PM" src="https://user-images.githubusercontent.com/5301572/155534967-d539c8e8-d280-4f25-b35a-dd4b3d1724fc.png">


I am trying to figure out why the two branches behave differently. This could be due to how they are merged in the SVN which could lead the `reposurgeon` to treat them differently.
So I would like to know how the merging in done on SVN currently?
@rjsparks @jennifer-richards Do you have any insights on this ?

**Comment by jennifer-richards** [comment]
I haven't found anything concrete, but maybe it depends on whether or not there were more commits on the personal branch after the one merged?

**Comment by rjsparks** [comment]
I've spotted the difference. From an SVN perspective the workflows here were essentially identical, but I can see reposurgeon possibly having a hard time.

Speaking in svn terms:
r19925 was a commit into personal/jennifer/7.44.1.dev0. r19932 was that commit being merged /trunk.
At https://trac.ietf.org/trac/ietfdb/changeset/19932 you can see:
![image](https://user-images.githubusercontent.com/10996692/155550641-86601983-e39a-4b94-9826-e541e8b93dea.png)
Note the svn:mergeinfo property

r19956 and r19947 followed the same workflow, and have similar svn properties, but r19947 was a commit to a node that was a subtree of /personal/rjs/7.45.1.dev0  (specifically /personal/rjs/7.45.1.dev0/ietf). 
*Edit*
![image](https://user-images.githubusercontent.com/10996692/155554375-6abe73d5-5a36-4ff7-9e84-200745b72d05.png)
*Edit ends*
The merge at r19956 merged the entire tree at that revision which picked up the the only changeset affecting the tree (r19947):
![image](https://user-images.githubusercontent.com/10996692/155552421-d9897524-2278-4f77-aa63-82564e41c109.png)
Again, note that /personal/rjs/7.45.1.dev0 was the source of the merge.

This kind of commit (to a subtree in a branch) has happened sporadically for reasons. It comes more frequently from volunteers working with older gui svn tools. In this particular case, I was the source of the subtree commit and it was a convenience to easily isolate some work that was going on elsewhere in the tree.

Let me know if this sparks an obvious "a-hah, we can easily adjust to deal with that". I would prefer to start the final conversion today, but if this is something that can be addressed in a small amount of time, I would like to see it fixed.

**Comment by rjsparks** [comment]
Sorry - long comment and I buried the lede. The essence is: 
> r19947 was a commit to a node that was a subtree of /personal/rjs/7.45.1.dev0

**Comment by rjsparks** [comment]
Jennifer pointed out that I jumped to an incorrect conclusion and that 19925 has the same kind of structure that 19947 had:
![image](https://user-images.githubusercontent.com/10996692/155561137-1dd7602e-14ad-4149-a61e-20380198fb68.png)
So, the above is all noise. 

My apologies, I'll keep looking to see if there really is a difference.

**Comment by jennifer-richards** [comment]
Just noticed r19967 (682392081bddbd1b8653df9135388e6b7c48ee1c) which was merged in r19968 (e3aa43eea55ea3fa4eadc557a1f148106438f8b9) - its merge has only one parent but has the same `personal/rjs/7.45.1.dev0/ietf` location.

I'm also not seeing a correlation with ends of branches being handled differently, so my guess was incorrect.

**Comment by rjsparks** [comment]
```
(env) rjsparks@undex trunk % svn propget svn:mergeinfo . | grep /personal/jennifer/7.44.1.dev0
/personal/jennifer/7.44.1.dev0:19895,19900,19917,19922,19925,19930
(env) rjsparks@undex trunk % svn propget svn:mergeinfo . | grep /personal/rjs/7.45.1.dev0
/personal/rjs/7.45.1.dev0:19946-19950,19952,19965,19967
```
Really - this is definitive, and fairly clear about what changesets came in.

**Comment by jennifer-richards** [comment]
So it seems that nothing on jennifer/7.44.1.dev0 was referenced when merged. Everything on rjs/7.45.1.dev0 was.

Commits from jennifer/7.45.1.dev0 branch seems to be referenced on merge, but I don't see anything obviously different in the creation of those branches ([r19893](https://trac.ietf.org/trac/ietfdb/changeset/19893) and [r19951](https://trac.ietf.org/trac/ietfdb/changeset/19951)).

**Comment by rjsparks** [comment]
@HassanAkbar @ronaldtse We can't see anything in svn that should have caused /personal/jennifer/7.44.1.dev0 to be treated differently than /personal/7.45.1.dev0. If you can, it would be nice to straighten it out. Otherwise, we can move on with what we have, even if it is suboptimal.

So, again, forgive the noisy discussion above, and if you see what to fix, please fix it, otherwise, please start the final conversion run for the datatracker and mailarchive and let me know when it's finished.

**Comment by HassanAkbar** [comment]
I’ve spent some time today on this by looking into internal `reposurgeon` code but I was unable to get to the root cause. This will take more time to investigate and fix.

So, I am starting the final migration for now.

**Comment by rjsparks** [comment]
@HassanAkbar - Please let us know when this is ready to transfer. But see #28 - I'd like to make sure we have the attachments in the right place for the datatracker and mailarchive.

**Comment by HassanAkbar** [comment]
While reviewing the migration for `datatracker` I found an issue with comment linkage. For keeping history whenever a field is changed in trac we add a comment in GitHub stating that change.

For example, in [this](https://github.com/ietf-svn-conversion/datatracker-old1/issues/2006#issuecomment-1053628435) issue comments are added like,
_@rjsparks@nostrum.com_ _changed status from `new` to `under_review`_

This changes the numbering of the comments so when adding the reference to a comment like `Replying to [comment:5]` it will link to wrong comment because of the additional comments added for property changes.

I’m looking for a solution to this, but it will take some time.
@rjsparks Is it okay to delay the migration for this?

**Comment by rjsparks** [comment]
If the links point to the right ticket, but just the wrong comment in the ticket, I think we can live with the damage, and would be inclined to move ahead (unless the fix would be _very_ quick).

**Comment by rjsparks** [comment]
@HassanAkbar If we move ahead without that fix, is the repository otherwise ready for us to transfer.

**Comment by HassanAkbar** [comment]
@rjsparks If we move ahead without the fix the `datatracker` and `mailarchive` repos will be ready by tomorrow, and if we fix the issue they will be ready by day after tomorrow.

**Comment by rjsparks** [comment]
Then we will wait for the fix and plan to move the repository on US Thursday. Please let me know when things are ready, or if something unexpected comes up.

**Comment by HassanAkbar** [comment]
> Then we will wait for the fix and plan to move the repository on US Thursday. Please let me know when things are ready, or if something unexpected comes up.

Sounds good 👍 
I will update you once the migration is done.

**Comment by rjsparks** [comment]
@HassanAkbar - we're ready to start, but need a signal from you that the repos are ready?

**Comment by HassanAkbar** [comment]
> @HassanAkbar - we're ready to start, but need a signal from you that the repos are ready?

@rjsparks 
The migration is still in progress. I am hoping to have it ready in around 5 hours from now. 
I will ping you right away as it gets completed.

**Comment by HassanAkbar** [comment]
@rjsparks 
Mailarch is ready > https://github.com/ietf-svn-conversion/mailarch

For the datatracker, everything else went good except there was a race condition faced when daily scheduled/automated migration cron started:

Issue# 712 was created twice and 713 was not created.
712 -> https://github.com/ietf-svn-conversion/datatracker-713/issues/712
713 -> https://github.com/ietf-svn-conversion/datatracker-713/issues/713 (should have been https://github.com/ietf-svn-conversion/datatracker-old/issues/713)

This is the only issue that I think I can fix by a single API call without running complete migration again.
But I don't have the `ietf-svn-bot` API token. Would it be possible to send me a token at `hassan.akbar@ribose.com` ?
Let me know either we can do that or we need to do a migration again?

Also, if we can wait for another day, `datatracker` alone migration won't take more than 16 hours.

**Comment by rjsparks** [comment]
I can send you the token.

**Comment by HassanAkbar** [comment]
Sounds good.
Btw I have also started another datatracker migration as a backup plan.

**Comment by HassanAkbar** [comment]
> Sounds good. Btw I have also started another datatracker migration as a backup plan.

Apologies, we won't be able to add past dates via a single API call.
I think we will have to wait for the new migration to complete.

**Comment by rjsparks** [comment]
ok - please advise when done.
We'll wait and move the mailarchive at the same time.

**Comment by HassanAkbar** [comment]
@rjsparks  Datatracker is also ready: https://github.com/ietf-svn-conversion/datatracker

**Comment by rjsparks** [comment]
Awesome! We're looking through them and expect to transfer them in the next couple of hours.

Could I get links to the changset->commithash maps for datatracker and mailarch?

**Comment by HassanAkbar** [comment]
Here are the link to svn number to Github commit hash file:
- datatracker -> https://github.com/ietf-ribose/svn-git-maps/blob/main/datatracker-revmap.txt
- mailarch -> https://github.com/ietf-ribose/svn-git-maps/blob/main/mailarch-revmap.txt

**Comment by HassanAkbar** [comment]
> 1. List of all page redirects that are required to ensure that external references to the existing Trac wikis are still functional. IETF will use this list to implement a redirect to the new page.
> 2. List of every Trac Wiki page migrated and its new target in GitHub Pages
> 3. List of every Trac Wiki page not migrated and why it was not moved.
> 4. Mappings from Subversion revision numbers to Git commit ids
> 5. Mappings from Trac ticket numbers to GitHub issue numbers.

1. [External URLs](https://github.com/ietf-ribose/github-migration-project/wiki/External-URLs)
2. [Migrated Wikis](https://github.com/ietf-ribose/github-migration-project/wiki/Migrated-Wikis#migrated)
3. [Skipped Wikis](https://github.com/ietf-ribose/github-migration-project/wiki/Migrated-Wikis#skipped), These are skipped because these are Trac internal wikis.
4. [This](https://github.com/ietf-ribose/svn-git-maps) repo contains all the mappings.
5. Trac issues and Github issues have 1-to-1 mapping.

@rjsparks Is there anything else required for the documentation?

**Comment by rjsparks** [comment]
You misunderstood what's needed for 1. What we need is the list (or algorithm) for redirecting URLs like this:
https://trac.ietf.org/trac/ietfdb/wiki/SprintCoderSetup -> https://github.com/ietf-tools/datatracker/wiki/SprintCoderSetup
if all pages map 1-1 with just the prefix change, that's all we need to know for 1).

(The question anticipated that some pages might have redirected to the mailarchive wiki, which we chose not to do).

**Comment by rjsparks** [comment]
Having what you prepared for 1 is still useful to us - it will inform future maintenance of the wiki.

**Comment by rjsparks** [comment]
At 4, what are the .fo files?

**Comment by rjsparks** [comment]
This is something we don't have yet (from the RFP/Contract):

> 7. Any links to existing Trac instances that appear in the RFC series, the Internet
Draft database or the IETF website, must be identified and documented. The
documentation must include where the target of each link has moved to. The
documentation must be provided, in a format that can easily be used to generate
a list of server redirections.

> 8. Any links to existing Subversion instances that appear in the RFC series, the
Internet Draft database or the IETF website, must be identified and documented.
The documentation must include where the target of each link has moved to.
The documentation must be provided, in a format that can easily be used to
generate a list of server redirections.

That may have been the original motivation for item 1 above.

**Comment by ronaldtse** [comment]
In 4. the `.fo` files are a mapping from SVN commits to a unique "action stamp" generated by reposurgeon (see "[Reference lifting](http://www.catb.org/~esr/reposurgeon/repository-editing.html#reference-lifting)" in the reposurgeon manual). 

This is a mechanism that allows someone to quickly look at the SVN revision number, locate the "action stamp", then in the converted Git repo to locate that commit using the "git alias" function documented in the link provided.

I've provided the code here for completeness:
```
[alias]
	# git stamp <commit-ish> - print a reposurgeon-style action stamp
	stamp = show -s --format='%cI!%ce'

	# git scommit <stamp> <rev-list-args> - list most recent commit that matches <stamp>.
	# Must also specify a branch to search or --all, after these arguments.
	scommit = "!f(){ d=${1%%!*}; a=${1##*!}; arg=\"--until=$d -1\"; if [ $a != $1 ]; then arg=\"$arg --committer=$a\"; fi; shift; git rev-list $arg ${1:+\"$@\"}; }; f"

	# git scommits <stamp> <rev-list-args> - as above, but list all matching commits.
	scommits = "!f(){ d=${1%%!*}; a=${1##*!}; arg=\"--until=$d --after $d\"; if [ $a != $1 ]; then arg=\"$arg --committer=$a\"; fi; shift; git rev-list $arg ${1:+\"$@\"}; }; f"

	# git smaster <stamp> - list most recent commit on master that matches <stamp>.
	smaster = "!f(){ git scommit \"$1\" master --first-parent; }; f"
	smasters = "!f(){ git scommits \"$1\" master --first-parent; }; f"

	# git shs <stamp> - show the commits on master that match <stamp>.
	shs = "!f(){ stamp=$(git smasters $1); shift; git show ${stamp:?not found} $*; }; f"

	# git slog <stamp> <log-args> - start git log at <stamp> on master
	slog = "!f(){ stamp=$(git smaster $1); shift; git log ${stamp:?not found} $*; }; f"

	# git sco <stamp> - check out most recent commit on master that matches <stamp>.
	sco = "!f(){ stamp=$(git smaster $1); shift; git checkout ${stamp:?not found} $*; }; f"
```

And the command can be used like this:
```sh
$ git stamp "2007-12-01T22:01:08Z!henrik@levkowetz.com"
```

**Comment by ronaldtse** [comment]
@rjsparks I've added the 7 (#31) and 8 (#32) tasks as part of the deliverable set.

For links on Datatracker, is it possible to provide some sort of grep results that match `{svn,trac}.ietf.org` or `{svn,trac}.tools.ietf.org`? We just don't have the data to work with...

**Comment by rjsparks** [comment]
It doesn't ask you to look for links on the datatracker itself. Just in the corpus of RFCs and Internet-Drafts, and the main IETF website. Finding references from RFCs and drafts should just be a grep as you describe above. I'll ask if there's a way to query the wagtail part of the website from inside wagtail.

But, if we have no manual mappings other than the svn changeset to git commit hash involved, the problem the requirement is trying to solve is straightforward.

**Comment by ronaldtse** [comment]
Thanks @rjsparks . I recall from one of our emails that we need to comb out links from the RFCs and Internet-Drafts, and while we've done that for RFCs, somehow from Google I thought there are more links. After investigating further, the number is actually feasible. We'll have the Internet-Draft links (from Datatracker) ready.

**Comment by ronaldtse** [comment]
#31 and #32 are both completed, and links from the datatracker have already been scraped into the deliverables of those tasks.

@rjsparks is there any remaining task we need to complete as part of the work under the RFP? Thanks.

**Comment by rjsparks** [comment]
I think this is very close.

It would be nice to have the various bits of documentation gathered into one easy spot for archiving to tie up this project.
A single repo with such things in it might be a good way to package things - at the moment, I find myself having to hunt.


Are the user mapping files in https://github.com/ietf-ribose/svn-github-convert what were actually used? I note that the datatracker user mapping file has a bunch of incorrect suggested maps in it. My email addresses are in there at least twice, with one pointing to a github user that is not me. No harm seems to have come of it. If these were the maps that were used, perhaps this could become part of the above suggested package.

Could what's at https://github.com/ietf-ribose/svn-git-maps go into that package?

**Comment by HassanAkbar** [comment]
@rjsparks [Here](https://github.com/ietf-ribose/Final-Deliverables) is the repo with all the stuff related to the migration. 

Let me know If you need anything else added to this repo.

**Comment by rjsparks** [comment]
migrated wikis is specific to ietfdb - we should say something about the working group wikis that were migrated. If you don't have a similar document already, a short paragraph saying that pages were migrated retaining the final path segment, but adding .md (as trac wiki content was converted to markdown), and that certain trac specific pages were not migrated.

**Comment by rjsparks** [comment]
(small suggestion - we have multiple projects in flight - maybe rename that repo to something specific to tracsvn)

**Comment by HassanAkbar** [comment]
> (small suggestion - we have multiple projects in flight - maybe rename that repo to something specific to tracsvn)

I've renamed the repo to [tracsvn-to-github-deliverables](https://github.com/ietf-ribose/tracsvn-to-github-deliverables)

> migrated wikis is specific to ietfdb - we should say something about the working group wikis that were migrated. If you don't have a similar document already, a short paragraph saying that pages were migrated retaining the final path segment, but adding .md (as trac wiki content was converted to markdown), and that certain trac specific pages were not migrated.

I've added a paragraph stating these changes at the start of the [migrated wikis](https://github.com/ietf-ribose/tracsvn-to-github-deliverables/blob/main/Migrated-Wikis.md) file.

Let me know If you need anything else added to this repo.

**Comment by rjsparks** [comment]
I think this deliverable complete, and modulo cleaning up artifacts we don't need to keep, I think the project is finished.
(@ronaldtse)

I'll leave the tracsvn-to-github-deliverables repo where it is until next week or so in case we find something else to put there. Then I'll move it to ietf-tools.

**Comment by ronaldtse** [comment]
That's perfect. Please feel free to move things when you see fit. Thank you @rjsparks for the support throughout!
